File: api/src/mappings/programmer.rs
use mizer_fixtures::definition::FixtureControlValue;
use protobuf::{EnumOrUnknown, MessageField};

use crate::models::*;

impl WriteControlRequest {
    pub fn as_controls(self) -> FixtureControlValue {
        use crate::models::fixtures::FixtureControl::*;

        match (self.control.unwrap(), self.value) {
            (INTENSITY, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Intensity(value)
            }
            (SHUTTER, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Shutter(value)
            }
            (FOCUS, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Focus(value)
            }
            (ZOOM, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Zoom(value)
            }
            (PRISM, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Iris(value)
            }
            (IRIS, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Iris(value)
            }
            (FROST, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Frost(value)
            }
            (PAN, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Pan(value)
            }
            (TILT, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Tilt(value)
            }
            (COLOR_MIXER, Some(write_control_request::Value::Color(value))) => {
                FixtureControlValue::ColorMixer(value.red, value.green, value.blue)
            }
            (COLOR_WHEEL, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::ColorWheel(value)
            }
            (GENERIC, Some(write_control_request::Value::Generic(value))) => {
                FixtureControlValue::Generic(value.name, value.value)
            }
            (GOBO, Some(write_control_request::Value::Fader(value))) => {
                FixtureControlValue::Gobo(value)
            }
            _ => unreachable!(),
        }
    }
}

impl From<mizer_fixtures::programmer::ProgrammerState> for ProgrammerState {
    fn from(state: mizer_fixtures::programmer::ProgrammerState) -> Self {
        Self {
            active_fixtures: state
                .active_fixtures
                .into_iter()
                .map(FixtureId::from)
                .collect(),
            fixtures: state
                .tracked_fixtures
                .into_iter()
                .map(FixtureId::from)
                .collect(),
            highlight: state.highlight,
            ..Default::default()
        }
    }
}

impl From<PresetId> for mizer_fixtures::programmer::PresetId {
    fn from(id: PresetId) -> Self {
        use preset_id::PresetType::*;
        match id.type_.unwrap() {
            Intensity => Self::Intensity(id.id),
            Shutter => Self::Shutter(id.id),
            Color => Self::Color(id.id),
            Position => Self::Position(id.id),
        }
    }
}

impl From<mizer_fixtures::programmer::PresetId> for PresetId {
    fn from(id: mizer_fixtures::programmer::PresetId) -> Self {
        use mizer_fixtures::programmer::PresetId::*;

        match id {
            Intensity(id) => Self {
                id,
                type_: EnumOrUnknown::new(preset_id::PresetType::Intensity),
                ..Default::default()
            },
            Shutter(id) => Self {
                id,
                type_: EnumOrUnknown::new(preset_id::PresetType::Shutter),
                ..Default::default()
            },
            Color(id) => Self {
                id,
                type_: EnumOrUnknown::new(preset_id::PresetType::Color),
                ..Default::default()
            },
            Position(id) => Self {
                id,
                type_: EnumOrUnknown::new(preset_id::PresetType::Position),
                ..Default::default()
            },
        }
    }
}

impl
    From<(
        mizer_fixtures::programmer::PresetId,
        mizer_fixtures::programmer::Preset<f64>,
    )> for Preset
{
    fn from(
        (id, preset): (
            mizer_fixtures::programmer::PresetId,
            mizer_fixtures::programmer::Preset<f64>,
        ),
    ) -> Self {
        Self {
            id: MessageField::some(id.into()),
            value: Some(preset::Value::Fader(preset.value)),
            label: preset.label,
            ..Default::default()
        }
    }
}

impl
    From<(
        mizer_fixtures::programmer::PresetId,
        mizer_fixtures::programmer::Preset<mizer_fixtures::programmer::Color>,
    )> for Preset
{
    fn from(
        (id, preset): (
            mizer_fixtures::programmer::PresetId,
            mizer_fixtures::programmer::Preset<mizer_fixtures::programmer::Color>,
        ),
    ) -> Self {
        Self {
            id: MessageField::some(id.into()),
            value: Some(preset::Value::Color(preset.value.into())),
            label: preset.label,
            ..Default::default()
        }
    }
}

impl From<mizer_fixtures::programmer::Color> for preset::Color {
    fn from((red, green, blue): mizer_fixtures::programmer::Color) -> Self {
        Self {
            red,
            green,
            blue,
            ..Default::default()
        }
    }
}

impl
    From<(
        mizer_fixtures::programmer::PresetId,
        mizer_fixtures::programmer::Preset<mizer_fixtures::programmer::Position>,
    )> for Preset
{
    fn from(
        (id, preset): (
            mizer_fixtures::programmer::PresetId,
            mizer_fixtures::programmer::Preset<mizer_fixtures::programmer::Position>,
        ),
    ) -> Self {
        Self {
            id: MessageField::some(id.into()),
            value: Some(preset::Value::Position(preset.value.into())),
            label: preset.label,
            ..Default::default()
        }
    }
}

impl From<mizer_fixtures::programmer::Position> for preset::Position {
    fn from((pan, tilt): mizer_fixtures::programmer::Position) -> Self {
        Self {
            pan,
            tilt,
            ..Default::default()
        }
    }
}

impl From<mizer_fixtures::programmer::Group> for Group {
    fn from(group: mizer_fixtures::programmer::Group) -> Self {
        Self {
            id: group.id,
            name: group.name,
            ..Default::default()
        }
    }
}

impl From<mizer_fixtures::programmer::ProgrammerChannel> for ProgrammerChannel {
    fn from(channel: mizer_fixtures::programmer::ProgrammerChannel) -> Self {
        use FixtureControlValue::*;
        let (control, value) = match channel.value {
            Intensity(value) => (
                FixtureControl::INTENSITY,
                programmer_channel::Value::Fader(value),
            ),
            Shutter(value) => (
                FixtureControl::SHUTTER,
                programmer_channel::Value::Fader(value),
            ),
            Pan(value) => (FixtureControl::PAN, programmer_channel::Value::Fader(value)),
            Tilt(value) => (
                FixtureControl::TILT,
                programmer_channel::Value::Fader(value),
            ),
            Focus(value) => (
                FixtureControl::FOCUS,
                programmer_channel::Value::Fader(value),
            ),
            Zoom(value) => (
                FixtureControl::ZOOM,
                programmer_channel::Value::Fader(value),
            ),
            Prism(value) => (
                FixtureControl::PRISM,
                programmer_channel::Value::Fader(value),
            ),
            Iris(value) => (
                FixtureControl::IRIS,
                programmer_channel::Value::Fader(value),
            ),
            Frost(value) => (
                FixtureControl::FROST,
                programmer_channel::Value::Fader(value),
            ),
            ColorMixer(red, green, blue) => (
                FixtureControl::COLOR_MIXER,
                programmer_channel::Value::Color(crate::models::fixtures::ColorMixerChannel {
                    red,
                    green,
                    blue,
                    ..Default::default()
                }),
            ),
            ColorWheel(value) => (
                FixtureControl::COLOR_WHEEL,
                programmer_channel::Value::Fader(value),
            ),
            Gobo(value) => (
                FixtureControl::GOBO,
                programmer_channel::Value::Fader(value),
            ),
            Generic(name, value) => (
                FixtureControl::GENERIC,
                programmer_channel::Value::Generic(programmer_channel::GenericValue {
                    value,
                    name,
                    ..Default::default()
                }),
            ),
        };

        Self {
            fixtures: channel.fixtures.into_iter().map(FixtureId::from).collect(),
            control: control.into(),
            value: Some(value),
            ..Default::default()
        }
    }
}

impl From<store_request::Mode> for mizer_command_executor::StoreMode {
    fn from(mode: store_request::Mode) -> Self {
        match mode {
            store_request::Mode::Merge => Self::Merge,
            store_request::Mode::AddCue => Self::AddCue,
            store_request::Mode::Overwrite => Self::Overwrite,
        }
    }
}

File: crates/ui/lib/api/plugin/ffi/programmer.dart
import 'dart:ffi' as ffi;

import 'package:ffi/ffi.dart';
import 'package:mizer/api/contracts/programmer.dart' show IProgrammerStatePointer;
import 'package:mizer/protos/fixtures.pb.dart';
import 'package:mizer/protos/programmer.pb.dart';

import 'package:mizer/api/plugin/ffi/bindings.dart';
import 'package:mizer/api/plugin/ffi/ffi_pointer.dart';

Map<int, FixtureControl> controlMappings = {
  FFIFixtureFaderControl.Intensity: FixtureControl.INTENSITY,
  FFIFixtureFaderControl.Shutter: FixtureControl.SHUTTER,
  FFIFixtureFaderControl.ColorMixer: FixtureControl.COLOR_MIXER,
  FFIFixtureFaderControl.ColorWheel: FixtureControl.COLOR_WHEEL,
  FFIFixtureFaderControl.Pan: FixtureControl.PAN,
  FFIFixtureFaderControl.Tilt: FixtureControl.TILT,
  FFIFixtureFaderControl.Focus: FixtureControl.FOCUS,
  FFIFixtureFaderControl.Zoom: FixtureControl.ZOOM,
  FFIFixtureFaderControl.Prism: FixtureControl.PRISM,
  FFIFixtureFaderControl.Iris: FixtureControl.IRIS,
  FFIFixtureFaderControl.Frost: FixtureControl.FROST,
  FFIFixtureFaderControl.Gobo: FixtureControl.GOBO,
};

class ProgrammerStatePointer extends FFIPointer<Programmer> implements IProgrammerStatePointer {
  final FFIBindings _bindings;

  ProgrammerStatePointer(this._bindings, ffi.Pointer<Programmer> ptr) : super(ptr);

  ProgrammerState readState() {
    var state = this._bindings.read_programmer_state(ptr);
    var activeFixtures = _readFixtureSelection(state.active_fixtures);
    var fixtures = _readFixtureSelection(state.fixtures);
    var selection = _readGroupedFixtureSelection(state.selection);
    var channels = _readProgrammerChannel(state.channels);
    var activeGroups = _readGroupSelection(state.active_groups);
    var effects = _readEffects(state.effects);

    return ProgrammerState(
      activeFixtures: activeFixtures,
      activeGroups: activeGroups,
      fixtures: fixtures,
      selection: selection,
      controls: channels,
      highlight: state.highlight == 1,
      offline: state.offline == 1,
      blockSize: state.block_size,
      groups: state.groups,
      wings: state.wings,
      effects: effects,
    );
  }

  List<FixtureId> _readFixtureSelection(Array_FFIFixtureId result) {
    var fixtures = new List.generate(result.len, (index) => result.array.elementAt(index).ref);

    return fixtures.map((id) => id.toFixtureId()).toList();
  }

  List<int> _readGroupSelection(Array_u32 result) {
    var groups = new List.generate(result.len, (index) => result.array.elementAt(index).value);

    return groups;
  }

  FixtureSelection _readGroupedFixtureSelection(Array_Array_FFIFixtureId result) {
    var groups = new List.generate(result.len, (index) => result.array.elementAt(index).ref);

    var fixtures = groups.map((fixtures) => _readFixtureSelection(fixtures)).toList();

    return FixtureSelection(
        fixtures: fixtures.map((f) => FixtureSelection_GroupedFixtureList(fixtures: f)).toList());
  }

  List<ProgrammerChannel> _readProgrammerChannel(Array_FFIProgrammerChannel result) {
    var channels = new List.generate(result.len, (index) => result.array.elementAt(index).ref);

    return channels.map((channel) {
      var fixtures = _readFixtureSelection(channel.fixtures);
      var control = controlMappings[channel.control];
      var result = ProgrammerChannel(
        control: control,
        fixtures: fixtures,
      );
      if (channel.preset == 1) {
        result.preset = _mapPresetId(channel.value.preset, control!);
      } else if (channel.control == FFIFixtureFaderControl.ColorMixer) {
        result.color = ColorMixerChannel(
          red: channel.value.color.red,
          green: channel.value.color.green,
          blue: channel.value.color.blue,
        );
      } else if (channel.control == FFIFixtureFaderControl.Generic) {
        result.generic = ProgrammerChannel_GenericValue(
          name: channel.value.generic.channel.cast<Utf8>().toDartString(),
          value: channel.value.generic.value,
        );
      } else {
        result.fader = channel.value.fader;
      }
      return result;
    }).toList();
  }

  List<EffectProgrammerState> _readEffects(Array_FFIEffectProgrammerState result) {
    var effects = new List.generate(result.len, (index) => result.array.elementAt(index).ref);

    return effects.map((effect) {
      return EffectProgrammerState(
        effectId: effect.effect_id,
        effectRate: effect.rate,
        effectOffset: effect.has_offset == 1 ? effect.effect_offset : null,
      );
    }).toList();
  }

  PresetId _mapPresetId(FFIPresetId presetId, FixtureControl control) {
    if (control == FixtureControl.INTENSITY) {
      return PresetId(id: presetId.intensity, type: PresetId_PresetType.INTENSITY);
    }
    if (control == FixtureControl.SHUTTER) {
      return PresetId(id: presetId.shutter, type: PresetId_PresetType.SHUTTER);
    }
    if (control == FixtureControl.PAN) {
      return PresetId(id: presetId.position, type: PresetId_PresetType.POSITION);
    }
    if (control == FixtureControl.COLOR_MIXER) {
      return PresetId(id: presetId.color, type: PresetId_PresetType.COLOR);
    }
    
    return PresetId();
  }
  
  @override
  void disposePointer(ffi.Pointer<Programmer> _ptr) {
    this._bindings.drop_programmer_pointer(_ptr);
  }
}

extension FFI on FFIFixtureId {
  FixtureId toFixtureId() {
    if (this.sub_fixture_id != 0) {
      return FixtureId(
          subFixture: SubFixtureId(fixtureId: this.fixture_id, childId: this.sub_fixture_id));
    }
    return FixtureId(fixture: this.fixture_id);
  }
}
